### ISSUE TITLE
Adding new frameworks to mesos

### ISSUE THREAD

**Issue by Y4Rv1K** [CiscoCloud/microservices-infrastructure#145]
Hello guys, few questions about adding extra frameworks to mesos. 

1. I'm understanding that major idea is using mesos with Marathon as core for other app  (hide complexity and provide service for apps) 

2. I see you already used ACL security list at mesos so every new framework required authentication. In such case if I want to add hdfs/spark as frameworks I should reconfigure your provision/bootstrap scripts appropriately (secret key generation, startup script for hdfs/spark/etc., etc.) or do it manually of course. 

In other words, if I want install hdfs and spark on top of mesos and I see no other solution rather change internal configuration for mesos and add few extra things.

Probably in this case I should go down to Mesos level and concentrate at frameworks itself. But that will be resulted in other project forked from this one. 

What will be proper way to extend this platform? Did I miss something?

**Comment by Y4Rv1K** [comment]
BTW Some current limitation related to this implementation and Mesos frameworks
[Framework authorisation at mesos](http://mesos.apache.org/documentation/latest/authorization/) is cool feature but looks like just a marathon [implemented it](https://mesosphere.github.io/marathon/docs/command-line-flags.html). Other frameworks still [working on it](https://issues.apache.org/jira/browse/SPARK-6284) (spark). 

That is additional topic to keep in mind.

**Comment by stevendborrelli** [comment]
I have some ideas around this, and I'd like to flesh out a design:

1) Folks will want to launch multiple frameworks on our cluster (Kafka, HDFS, Cassandra, elastic search, etc.)
2) We require framework auth, so every new framework will need to add credentials to the leader's auth files
3) We'll want to use Marathon to run the framework process

What I am thinking is we develop a couple of things:
1) Consul-template to manage framework and follower auth credentials. This would be similar to the ngninx auth template. There would be 2 consul k/v endpoints:
   `/mesos/auth/frameworks/<framework>=credential`
   `/mesos/auth/follower/<follower>=credential`
That way we can just use a JSON put to add new credentials to all the leader nodes without having to re-run ansible. 

3) create a framework launcher. The framework launcher  would be a simple web app running in marathon that wraps the marathon API. 

We could create a ready-set number of marathon templates and make them public in github, so that all the user had to do was run:
    ```deploy kafka```
    ```deploy hdfs```

And we'd take care of credentials, port mapping, nginx proxying, etc.

**Comment by BrianHicks** [comment]
For now if you want to use a framework which doesn't support authentication, you can set `mesos_authenticate_frameworks` to a falsey value. ([doc](https://microservices-infrastructure.readthedocs.org/en/latest/components/mesos.html#mesos_authenticate_frameworks))

**Comment by Y4Rv1K** [comment]
Thanks @BrianHicks 

>> create a framework launcher. The framework launcher would be a simple web app running in marathon that wraps the marathon API.

@stevendborrelli looks like it's best way to do it.

**Comment by OleksandrBerchenko** [comment]
It seems that "mesos_authenticate_frameworks" option doesn't work.

I got the following error when running Spark shell with Mesos:

$ ./bin/spark-shell --master mesos://host-02:2181
...
I0414 15:46:10.015877 31460 sched.cpp:157] Version: 0.22.0
I0414 15:46:10.019696 31461 sched.cpp:254] New master detected at master@10.203.50.202:2181
I0414 15:46:10.019978 31461 sched.cpp:264] No credentials provided. Attempting to register without authentication

(after that Spark shell hangs).

Ok, that's probably an expected behavior taking into account Mesos authentication.  So, I decided to disable it:

1. Edit roles/mesos/defaults/main.yml and set:

mesos_authenticate_frameworks: False

2. Re-run site.yml (I didn't create a cluster from scratch):

$ ansible-playbook -i inventory/1-datacenter site.yml -e @security.yml

3. Try to run Spark shell one more time:

$ ./bin/spark-shell --master mesos://host-02:2181

The result was the same: it hangs after "No credentials provided. Attempting to register without authentication" message.

Do I miss anything?

Thanks!

**Comment by stevendborrelli** [comment]
Maybe there is a port issue? 

2181 is the zookeeper port, mesos is port 5050. 

Can you try --master mesos://host-02:5050?

We might run into other authentication problems, but I want to see if that changes anything.

**Comment by OleksandrBerchenko** [comment]
No, this didn't help: the result is exactly the same.

Please note, that when I tried:

$ ./bin/spark-shell --master mesos://zk://host-02:2181

as http://spark.apache.org/docs/latest/running-on-mesos.html suggests, I got the following error:

15/04/14 18:21:07 ERROR MesosSchedulerBackend: Mesos error: Failed to create a master detector for 'zk://host-02:2181': Expecting a (chroot) path for ZooKeeper ('/' is not supported)
15/04/14 18:21:07 ERROR TaskSchedulerImpl: Exiting due to error from cluster scheduler: Failed to create a master detector for 'zk://host-02:2181': Expecting a (chroot) path for ZooKeeper ('/' is not supported)

Thanks!

**Comment by stevendborrelli** [comment]
@OleksandrBerchenko I'm concerned that 
`./bin/spark-shell --master mesos://zk://host-02:2181` is the wrong format. 

it should be `--master mesos://host-02:5050`. Can you try that and report back?

If that still fails, on host-02 mesos will log in `/var/log/mesos`. There are several master logs there.

**Comment by ChrisAubuchon** [comment]
That looks like it's expecting `/mesos` at the end of the master string:
Something like:
```
mesos://zk://host-02:2181/mesos
```

**Comment by stevendborrelli** [comment]
@OleksandrBerchenko I'm testing with spark to see what the issue is. 
These lines below are actually coming from mesos itself: 

```
I0414 15:46:10.015877 31460 sched.cpp:157] Version: 0.22.0
I0414 15:46:10.019696 31461 sched.cpp:254] New master detected at master@10.203.50.202:2181
I0414 15:46:10.019978 31461 sched.cpp:264] No credentials provided. Attempting to register without authentication
```

**Comment by stevendborrelli** [comment]
@OleksandrBerchenko Can you try the following command?

```
./spark-shell --master mesos://zk://zookeeper.service.consul:2181/mesos
```

**Comment by OleksandrBerchenko** [comment]
Well...

1. mesos://host-02:5050 doesn't work.
2. Both mesos://zk://host-02:2181/mesos and mesos://zk://zookeeper.service.consul:2181/mesos work inside the cluster.
3. When I use mesos://zk://host-02:2181/mesos outside of the cluster I get the same error:

No credentials provided. Attempting to register without authentication

4. When I use mesos://zk://zookeeper.service.consul:2181/mesos outside of the cluster I get a new error:

```
F0415 16:12:12.815044  2106 zookeeper.cpp:113] Failed to create ZooKeeper, zookeeper_init: No such file or directory [2]
*** Check failure stack trace: ***
    @     0x7fa092af28be  google::LogMessage::Fail()
    @     0x7fa092af281d  google::LogMessage::SendToLog()
    @     0x7fa092af222e  google::LogMessage::Flush()
    @     0x7fa092af2043  google::LogMessage::~LogMessage()
    @     0x7fa092af2f72  google::ErrnoLogMessage::~ErrnoLogMessage()
    @     0x7fa0926792b0  ZooKeeperProcess::initialize()
    @     0x7fa092a62c60  process::ProcessManager::resume()
    @     0x7fa092a5856e  process::schedule()
    @     0x7fa1accd2df3  start_thread
    @     0x7fa1ac5ee1ad  __clone
./bin/spark-shell: line 55:  1605 Aborted                 (core dumped) "$FWDIR"/bin/spark-submit --class org.apache.spark.repl.Main "${SUBMISSION_OPTS[@]}" spark-shell "${APPLICATION_OPTS[@]}"
```

Any suggestions how to make it working outside of the cluster?  I have already opened port 2181.  What else should I do?

Thanks!

**Comment by stevendborrelli** [comment]
@OleksandrBerchenko 

Currently mesos listens on port 15050, with an nginx proxy in front at 5050. 

There is an iptables firewall that only allows connections from other mesos systems in the cluster to port 15050. You can allow access to port 15050 if you modify the MESOSLEADER iptables chain.

**Comment by ChrisAubuchon** [comment]
The 5050 requires password authentication from outside the cluster. Can you try:
```
mesos://admin:{admin password}@host-02:5050
```

**Comment by OleksandrBerchenko** [comment]
> mesos://admin:{admin password}@host-02:5050

```
15/04/17 17:52:44 ERROR MesosSchedulerBackend: Mesos error: Failed to create a master detector for 'admin:admin@rspark-master-01:5050': Failed to parse 'admin:admin@rspark-master-01:5050'
15/04/17 17:52:44 ERROR TaskSchedulerImpl: Exiting due to error from cluster scheduler: Failed to create a master detector for 'admin:admin@rspark-master-01:5050': Failed to parse 'admin:admin@rspark-master-01:5050'
```

**Comment by cdmitri** [comment]
I've got it to register the framework, not a priority right now, let's get
the basic role out first.

**Comment by OleksandrBerchenko** [comment]
> There is an iptables firewall that only allows connections from other mesos systems in the cluster to port 15050. You can allow access to port 15050 if you modify the MESOSLEADER iptables chain.

Well, I have updated roles/mesos/templates/mesos-leader-iptables.tmpl.j2 changing:

```
iptables -A MESOSLEADER -p tcp --dport {{ mesos_leader_port }} -j DROP
```

to

```
iptables -A MESOSLEADER -p tcp --dport {{ mesos_leader_port }} -j ACCEPT
```

After this change I'm finally able to make a telnet collection to port 15050 from outside the cluster:

Unfortunately, that didn't help to resolve the issue with Spark.  It still hangs on this step:

```
I0423 15:16:11.273902  3691 detector.cpp:452] A new leading master (UPID=master@173.39.243.160:15050) is detected
I0423 15:16:11.274780  3688 sched.cpp:254] New master detected at master@173.39.243.160:15050
I0423 15:16:11.277977  3688 sched.cpp:264] No credentials provided. Attempting to register without authentication
```

Is there anything else that should be done?

Thanks!

**Comment by cdmitri** [comment]
@OleksandrBerchenko please work with Timofii @tymofii-polekhin and https://github.com/CiscoCloud/microservices-infrastructure/issues/176 to get the debugging done for this

**Comment by BrianHicks** [comment]
I see a number of issues still here, but to track the work more accurately I'm going to close the ticket (I believe the original question is resolved. If it isn't, please reopen.) If you feel that an issue you raised here is still outstanding and not addressed by an existing ticket, please open a new one. Thanks!
